platform-tests/src/test/java/org/junit/platform/engine/support/descriptor/ResourceLockAnnotationTests.java

```diff
@@ -0,0 +1,377 @@
+				NoSharedResourcesTestCase.NestedClass.class
+		);
+		assertThat(nestedClassResources).isEmpty();
+		// @formatter:on
+	}
+
+	@Test
+	void addSharedResourcesViaAnnotationValue() {
+		// @formatter:off
+		var classResources = getClassResources(
+				SharedResourcesViaAnnotationValueTestCase.class
+		);
+		assertThat(classResources).containsExactlyInAnyOrder(
+				new ExclusiveResource("a1", LockMode.READ_WRITE),
+				new ExclusiveResource("a2", LockMode.READ_WRITE)
+		);
+
+		var methodResources = getMethodResources(
+				SharedResourcesViaAnnotationValueTestCase.class
+		);
+		assertThat(methodResources).containsExactlyInAnyOrder(
+				new ExclusiveResource("b1", LockMode.READ),
+				new ExclusiveResource("b2", LockMode.READ_WRITE)
+		);
+
+		var nestedClassResources = getNestedClassResources(
+				SharedResourcesViaAnnotationValueTestCase.NestedClass.class
+		);
+		assertThat(nestedClassResources).containsExactlyInAnyOrder(
+				new ExclusiveResource("c1", LockMode.READ),
+				new ExclusiveResource("c2", LockMode.READ)
+		);
+		// @formatter:on
+	}
+
+	@Test
+	void addSharedResourcesViaAnnotationProviders() {
+		// @formatter:off
+		var classResources = getClassResources(
+				SharedResourcesViaAnnotationProvidersTestCase.class
+		);
+		assertThat(classResources).containsExactlyInAnyOrder(
+				new ExclusiveResource("a1", LockMode.READ),
+				new ExclusiveResource("a2", LockMode.READ)
+		);
+
+		var methodResources = getMethodResources(
+				SharedResourcesViaAnnotationProvidersTestCase.class
+		);
+		assertThat(methodResources).containsExactlyInAnyOrder(
+				new ExclusiveResource("b1", LockMode.READ_WRITE),
+				new ExclusiveResource("b2", LockMode.READ_WRITE)
+		);
+
+		var nestedClassResources = getNestedClassResources(
+				SharedResourcesViaAnnotationProvidersTestCase.NestedClass.class
+		);
+		assertThat(nestedClassResources).containsExactlyInAnyOrder(
+				new ExclusiveResource("c1", LockMode.READ_WRITE),
+				new ExclusiveResource("c2", LockMode.READ)
+		);
+		// @formatter:on
+	}
+
+	@Test
+	void addSharedResourcesViaAnnotationValueAndProviders() {
+		// @formatter:off
+		var classResources = getClassResources(
+				SharedResourcesViaAnnotationValueAndProvidersTestCase.class
+		);
+		assertThat(classResources).containsExactlyInAnyOrder(
+				new ExclusiveResource("a1", LockMode.READ_WRITE),
+				new ExclusiveResource("a2", LockMode.READ)
+		);
+
+		var methodResources = getMethodResources(
+				SharedResourcesViaAnnotationValueAndProvidersTestCase.class
+		);
+		assertThat(methodResources).containsExactlyInAnyOrder(
+				new ExclusiveResource("b1", LockMode.READ),
+				new ExclusiveResource("b2", LockMode.READ)
+		);
+
+		var nestedClassResources = getNestedClassResources(
+				SharedResourcesViaAnnotationValueAndProvidersTestCase.NestedClass.class
+		);
+		assertThat(nestedClassResources).containsExactlyInAnyOrder(
+				new ExclusiveResource("c1", LockMode.READ_WRITE),
+				new ExclusiveResource("c2", LockMode.READ_WRITE)
+		);
```

Comment:
The provider declared on the outermost class is called as well, right? I think we should change the test to show that.

Reply:
Yes, that's right.
But such case is already covered in `ResourceLocksProviderTests#ClassLevelProviderTestCase`.
Do you think we should cover it here as well?

Reply:
Indeed. We didn't test such case.
Added